### PR TITLE
daemon: Teach eos-event-recorder-daemon to upload metrics in new wire format

### PR DESCRIPTION
**Pull request by starnight**
* Remove the machine ID and send number from the payload
* Add the image ID and dictionary of location site ID data to the payload
* Add the boot type to the payload
* Remove the unix user ID field from each singular event payload
* Add current OS version into singular event payload
* Empty the array of aggregate events and update its string format
* Drop sequence events' actions by making it as no-operation

https://phabricator.endlessm.com/T32144

### CODE REVIEW

**Review by dbnicholson — CHANGES_REQUESTED** [review]
This looks good overall if I'm reading @wjt's comments from https://phabricator.endlessm.com/T32144 correctly. I left a few comments, but the direction seems correct.

It would be nice if some of the properties that don't change after boot could be gathered once, but I think what you have is probably good for a start.

File: daemon/emer-image-id-provider.h
Comment (starnight) [reply daemon/emer-image-id-provider.h:30]: sure!

File: daemon/emer-image-id-provider.c
Comment (starnight) [reply daemon/emer-image-id-provider.c:28]: Okay!  Will modify the comment.

File: tests/daemon/test-daemon.c
Comment (starnight) [reply tests/daemon/test-daemon.c:661]: Sure!  Will add an assert here.

File: daemon/emer-site-id-provider.c
Comment (starnight) [reply daemon/emer-site-id-provider.c:27]: Sure!

File: daemon/emer-site-id-provider.h
Comment (starnight) [reply daemon/emer-site-id-provider.h:30]: Sure!

File: tests/daemon/test-daemon.c
Comment (starnight) [reply tests/daemon/test-daemon.c:655]: Try to have the child first:
```
+  gsize num_siteid_values = g_variant_n_children (site_id);
+  g_info ("%s: site_id has %ld children", __func__, num_siteid_values);
+  for (gsize i = 0; i < num_siteid_values; i++)
+    {
+      gchar *key, *val;
+      g_variant_get_child(site_id, i, "{ss}", &key, &val);
+      g_info ("%s: {%s, %s}", __func__, key, val);
+    }
+  g_variant_unref (site_id);
```
The weird things are:
* **site_id** has no children!???
*  **site_id** cannot be unreferenced.  It shows error `Bail out! GLib-FATAL-CRITICAL: g_atomic_ref_count_dec: assertion 'g_atomic_int_get (arc) > 0' failed`.

Comment (starnight) [reply tests/daemon/test-daemon.c:655]: Also, checked `create_request_body()`'s [**site_id**](https://github.com/endlessm/eos-event-recorder-daemon/commit/de50c7bac8abba911c9a6ccb787e660bab1c28a3#diff-fadc9fc0899639577ac993bdc978725b109a6ec299404b9eb2817eb3df59c972R851) in `daemon/emer-daemon.c`.

It does get correct content:`[{"id", "myid"}, {"country", "Earth"}]` from [`emer_read_location_label()`](https://github.com/endlessm/eos-event-recorder-daemon/commit/de50c7bac8abba911c9a6ccb787e660bab1c28a3#diff-b9a781d12fdd6a794c150bb50aade43c8c6a9288ed50a8c2286e6e910ee3a6baR27) invoked by [`emer_get_site_id()`](https://github.com/endlessm/eos-event-recorder-daemon/commit/de50c7bac8abba911c9a6ccb787e660bab1c28a3#diff-b9a781d12fdd6a794c150bb50aade43c8c6a9288ed50a8c2286e6e910ee3a6baR40) in `tests/daemon/mock-site-id-provider.c`.

File: tests/daemon/test-daemon.c
Comment (starnight) [reply tests/daemon/test-daemon.c:434]: I think this is due to the [OS version string](https://github.com/endlessm/eos-event-recorder-daemon/commit/d5e1baac61dfb4738e2f5bd0a4e90c2dbf770701#diff-422d7470ef37a9b6f93e9098cb8b29c282c73ea5f4d965bb4b01f139fad3d60cR523), whose length is not fixed and variant from different OS.

File: tests/daemon/test-daemon.c
Comment (starnight) [reply tests/daemon/test-daemon.c:534]: Sure!

File: tests/daemon/test-daemon.c
Comment (starnight) [reply tests/daemon/test-daemon.c:739]: Sure!  Will add the comment here.

File: daemon/emer-boot-id-provider.h
Comment (starnight) [reply daemon/emer-boot-id-provider.h:89]: Sure!

File: daemon/emer-boot-id-provider.c
Comment (starnight) [reply daemon/emer-boot-id-provider.c:300]: The original [`EmerBootIdProviderPrivate`](https://github.com/endlessm/eos-event-recorder-daemon/blob/edcb6abcccc33aaf33d1908065d71deb318c3055/daemon/emer-persistent-cache.c#L53) is only used/related by `EmerPersistentCachePrivate` in `daemon/emer-persistent-cache.c`, not `daemon/emer-daemon.c`.  It is hard and complicated to get the **boot_type** via `daemon/emer-persistent-cache.c` in `daemon/emer-daemon.c`.

The **boot_type** is good to be cached, but is it good to be a property in `EmerBootIdProviderPrivate`?

File: daemon/emer-daemon.c
Comment (starnight) [reply daemon/emer-daemon.c]: Sure

File: daemon/emer-boot-id-provider.c
Comment (dbnicholson) [reply daemon/emer-boot-id-provider.c:300]: Not needed here, but I think what you'd do is add an instance of `EmerBootIdProvider` to `EmerDaemonPrivate` like the other providers. Then you would use the `get_boot_type` method during the request construction. In the `get_boot_type` implementation you'd use a `static` variable like in `emer_boot_id_provider_get_id` so that you only read it once store the result in `EmerBootIdProviderPrivate`.

For the persistent cache you could just have another instance of `EmerBootIdProvider` or you could pass it in when calling `emer_persistent_cache_new` from `emer_daemon_constructed`. `emer_persistent_cache_new_full` even supports passing in the boot ID provider already.

But anyways it's a bit premature optimization.

File: tests/daemon/test-daemon.c
Comment (dbnicholson) [reply tests/daemon/test-daemon.c:656]: GVariant always confuses me. The issue is that `site_id` is referring to a dictionary `a{ss}`. In that case, per the [dictionaries documentation](https://developer.gnome.org/glib/stable/gvariant-format-strings.html#gvariant-format-strings-dictionaries), the format passed to `g_variant_get` should be `@a{ss}` to refer to a pointer to a `GVariant`. Then per the [GVariant *](https://developer.gnome.org/glib/stable/gvariant-format-strings.html#gvariant-format-strings-gvariant) documentation, this will set the pointer to a reference, so (I believe) it should be dereferenced when done.

The following worked for me:
```
diff --git a/tests/daemon/test-daemon.c b/tests/daemon/test-daemon.c
index dce5e30..73d10d9 100644
--- a/tests/daemon/test-daemon.c
+++ b/tests/daemon/test-daemon.c
@@ -655,11 +655,23 @@ get_events_from_request (GByteArray    *request,
   GVariant *site_id;
   guchar boot_type;
   g_variant_get (native_endian_request,
-                 "(xxsa{ss}ya(aysxmv)a(ayssxmv))",
+                 "(xxs@a{ss}ya(aysxmv)a(ayssxmv))",
                  &client_relative_time, &client_absolute_time, &image_version,
                 &site_id, &boot_type, singular_iterator, aggregate_iterator);
   g_free (image_version);
 
+  const gchar *site_value;
+  gboolean site_ret;
+  g_assert_nonnull (site_id);
+  g_assert_cmpint (g_variant_n_children (site_id), ==, 2);
+  site_ret = g_variant_lookup (site_id, "id", "&s", &site_value);
+  g_assert_true (site_ret);
+  g_assert_cmpstr (site_value, ==, "myid");
+  site_ret = g_variant_lookup (site_id, "country", "&s", &site_value);
+  g_assert_true (site_ret);
+  g_assert_cmpstr (site_value, ==, "Earth");
+  g_variant_unref (site_id);
+
   g_assert_cmpint (client_relative_time, >=, fixture->relative_time);
   g_assert_cmpint (client_relative_time, <=, curr_relative_time);
 
```

File: tests/daemon/test-daemon.c
Comment (dbnicholson) [reply tests/daemon/test-daemon.c:435]: Right. I guess you'd need to mock out that part to return a fixed string in the test. In this particular commit it doesn't need to change (you've only removed a single byte), but later it does need to change when adding the OS version to singular events. Not worth doing anything about right now.

Comment (starnight) [reply tests/daemon/test-daemon.c:435]: Mock OS version string.   Hmm ...
It makes the behavior stable.  Make sense.  Sounds good!

File: tests/daemon/test-daemon.c
Comment (starnight) [reply tests/daemon/test-daemon.c:655]: Uh!  The '@' character again!  It always confuses me.  Thanks @dbnicholson

File: daemon/emer-boot-id-provider.c
Comment (starnight) [reply daemon/emer-boot-id-provider.c:300]: Okay!  Thanks

File: daemon/emer-boot-id-provider.c
Comment (dbnicholson) [inline daemon/emer-boot-id-provider.c:304]: Can you also make this `guint8` to match the return type?

File: daemon/emer-daemon.c
Comment (dbnicholson) [inline daemon/emer-daemon.c:435]: `s/guchar/guint8/`

File: tests/daemon/test-daemon.c
Comment (dbnicholson) [inline tests/daemon/test-daemon.c:734]: Can you remove this trailing newline?

**Review by dbnicholson — CHANGES_REQUESTED** [review]
Looking good. I just left a couple nitpicks I missed the first time. However, I think we should block merging this until endlessm/azafea#155 lands since otherwise we'll just log a bunch of errors sending metrics somewhere no one is listening.

File: tests/daemon/test-daemon.c
Comment (starnight) [reply tests/daemon/test-daemon.c]: sure!

File: daemon/emer-boot-id-provider.c
Comment (starnight) [reply daemon/emer-boot-id-provider.c]: ok

File: tests/daemon/test-daemon.c
Comment (starnight) [reply tests/daemon/test-daemon.c]: ok

File: daemon/emer-daemon.c
Comment (starnight) [reply daemon/emer-daemon.c]: ok

File: tests/daemon/test-daemon.c
Comment (starnight) [reply tests/daemon/test-daemon.c]: ok

**Review by dbnicholson — APPROVED** [review]
Looks good! As stated before, I'm going to leave this unmerged until the server side is accepting the v3 protocol.

**Comment by wjt** [comment]
I have deployed the server changes to the test environment. I think it is OK to merge this as-is. I don't think it will cause errors even in the dev server environment (which is what all master systems submit to by default) because azafea-metrics-proxy was updated to add the `v3` endpoint months ago, and it will just happily stick stuff onto the v3 redis queue, where they will wait until azafea is updated to pull from there.

**Pull request merged**
